lib/Backend/Lower.cpp

```diff
@@ -3816,14 +3813,155 @@ Lowerer::GenerateArrayAlloc(IR::Instr *instr, uint32 * psize, Js::ArrayCallSiteI
+        PUSH recycler
+        PUSH alignedArrayAllocSize
+        s10 = CALL AllocMemForJavascriptIntArray
+
+        ; ***** Load headSeg/initialize it *****
+        dstOpnd                 = s10
+        headOpnd                = LEA [s11+36]
+
+        ; ***** Evaluate arrayFlags *****
+        CMP sizeOpnd,           0
+        JE  $noMissingValues
+        MOV arrayFlagsOpnd,     Js::DynamicObjectFlags::ObjectArrayFlagsTag
+        JMP $skip
+$noMissingValues:
+        MOV arrayFlagsOpnd,     Js::DynamicObjectFlags::InitialArrayValue
+$skip:
+        ; ***** Array object initialization *****
+        ...
+        ...
+        MOV [dstOpnd+12],       arrayFlagsOpnd                  ; set arrayFlags
+        MOV [dstOpnd+16],       arrayLenOpnd                    ; set length field of JavascriptNativeIntArray
+        ...
+        MOV [headOpnd+4],       arrayLenOpnd                    ; set length field of HeadSegment
+        MOV [headOpnd+4],       alignedHeadSegSize              ; set size field of HeadSegment
+    */
+
+    IR::Opnd * sizeOpnd = IR::RegOpnd::New(TyUint32, func);
+    IR::Opnd * arraySizeOpnd = IR::RegOpnd::New(TyUint32, func);
+    
+    IR::Opnd * alignedArrayAllocSizeOpnd = IR::RegOpnd::New(TyUint32, func);
+    IR::Opnd * tempOpnd1 = IR::RegOpnd::New(TyUint32, func);
+    IR::Opnd * tempOpnd2 = IR::RegOpnd::New(TyUint32, func);
+    *alignedHeadSegSizeOpnd = IR::RegOpnd::New(TyUint32, func);
+
+    // ***** size = (arrayLen == 0) ? Js::SparseArraySegmentBase::HEAD_CHUNK_SIZE ? arrayLen *****
+    Lowerer::InsertMove(sizeOpnd, arrayLenOpnd, instr);
+    Lowerer::InsertCompare(sizeOpnd, IR::IntConstOpnd::New(0, TyUint32, func), instr);
+
+    IR::LabelInstr * notSmallChunkSizeLabel = IR::LabelInstr::New(Js::OpCode::Label, func);
+    Lowerer::InsertBranch(Js::OpCode::BrGt_A, notSmallChunkSizeLabel, instr);
+    Lowerer::InsertMove(sizeOpnd, IR::IntConstOpnd::New(Js::SparseArraySegmentBase::HEAD_CHUNK_SIZE, TyUint32, func), instr);
+    instr->InsertBefore(notSmallChunkSizeLabel);
+
+    // ***** calculate alignedArrayAllocationSize and alignedHeadSegSize *****
+    IR::Opnd * objectSizeOpnd = IR::IntConstOpnd::New(Js::JavascriptArray::DetermineObjectSize<ArrayType, 0>(), TyUint32, func);
+
+    // Instead of performing (arraySize * sizeof(TElement), we would use shift operation instead.
+    // The multiplier operand calculates by how much we should shift arraySize to left to achieve the same result.
+    IR::Opnd * sizeOfMultiplierOpnd = IR::IntConstOpnd::New(Math::Log2(sizeof(typename ArrayType::TElement)), TyUint32, func);
+    Lowerer::InsertShift(Js::OpCode::Shl_A, false, arraySizeOpnd, sizeOpnd, sizeOfMultiplierOpnd, instr);
+    Lowerer::InsertAdd(false, tempOpnd1, objectSizeOpnd, arraySizeOpnd, instr);
+
+#if defined(_M_IX86_OR_ARM32)
+    Lowerer::InsertAdd(false, tempOpnd2, tempOpnd1, IR::IntConstOpnd::New(HeapConstants::ObjectGranularity - 1, TyUint32, func), instr);
```

Comment:
>HeapConstants::ObjectGranularity - 1, TyUint32 [](start = 74, length = 46)

You can use TyVar. and for ObjectGranularity -> you can cast it to IntConstType ?
So you don't need a ifdef here.